xtext.cpp

```diff
@@ -0,0 +1,2157 @@
+                for ( int i=start; i<=end; i++ ) {
+                    if ( i > t_start && (i == end || m_charinfo[i].flags & CHAR_SCRIPT_CHANGE) ) {
+                        int hints = 0;
+                        if ( m_charinfo[t_start].flags & CHAR_IS_PARA_START )
+                            hints |= HINT_BEGINS_PARAGRAPH;
+                        if ( m_charinfo[i-1].flags & CHAR_IS_PARA_END )
+                            hints |= HINT_ENDS_PARAGRAPH;
+                        shapeSegment(0, t_start, i, hints, nb_glyphs);
+                        t_start = i;
+                    }
+                }
+            }
+            else { // Simple, only one LTR segment
+                int hints = 0;
+                if ( m_charinfo[start].flags & CHAR_IS_PARA_START )
+                    hints |= HINT_BEGINS_PARAGRAPH;
+                if ( m_charinfo[end-1].flags & CHAR_IS_PARA_END )
+                    hints |= HINT_ENDS_PARAGRAPH;
+                shapeSegment(0, start, end, hints, nb_glyphs);
+            }
+        }
+
+        // todo: do something about spaces at end? may be, just mark them
+        // and provide a width_without_trailing_spaces, so text justification
+        // can ignore them.
+
+        // Restore the char that we replaced with an ellipsis
+        if ( orig_idx_substituted ) {
+            m_text[idx_to_substitute_with_ellipsis] = orig_idx_char;
+            m_charinfo[idx_to_substitute_with_ellipsis].flags = orig_idx_charinfo_flags;
+        }
+
+        // Convert out s_shape_result to a Lua array.
+        // We will add some global metrics as table keys/values.
+        int total_advance = 0;
+        int nb_can_extend = 0;
+
+        lua_createtable(m_L, nb_glyphs, 0 ); // array of glyphs, pre-sized
+        for(int i = 0; i < nb_glyphs; i++) {
+            xtext_shapeinfo_t * s = &s_shape_result[i];
+
+            total_advance += s->x_advance;
+            if (s->can_extend)
+                nb_can_extend++;
+
+            lua_newtable(m_L); // key/value table of info about a single glyph
+
+            lua_pushstring(m_L, "font_num");
+            lua_pushinteger(m_L, s->font_num);
+            lua_settable(m_L, -3);
+
+            lua_pushstring(m_L, "glyph");
+            lua_pushinteger(m_L, s->glyph);
+            lua_settable(m_L, -3);
+
+            lua_pushstring(m_L, "text_index");
+            lua_pushinteger(m_L, s->text_index + 1); // (Lua indices start at 1)
+            lua_settable(m_L, -3);
+
+            lua_pushstring(m_L, "x_advance");
+            lua_pushinteger(m_L, s->x_advance);
+            lua_settable(m_L, -3);
+
+            lua_pushstring(m_L, "x_offset");
+            lua_pushinteger(m_L, s->x_offset);
+            lua_settable(m_L, -3);
+
+            lua_pushstring(m_L, "y_offset");
+            lua_pushinteger(m_L, s->y_offset);
+            lua_settable(m_L, -3);
+
+            lua_pushstring(m_L, "is_rtl");
+            lua_pushboolean(m_L, s->is_rtl);
+           
```

Comment:
s/indice/index/

(index is singular, indices is plural. indexes is accepted, but I hate it ;p).

Reply:
(Random comment: that's why `i` or `idx` is often used in variable names that represent an index ;)).

Reply:
GitHub looks a bit drunk (I assume it's github and not you, because it also does me strange things), but all your last comments points to that bit of code about "is_rtl" :) But I'll find out where they belong, no worry :)

Reply:
Oh, yeah, the recap in the "Conversation" panel inlines completely unrelated code (or, well, at least that certainly wasn't where I put 'em ;p).

Reply:
@poire-z: Appears to point to the right place in the "Changes" panel on my end ;).